agent/agent.go

```diff
@@ -340,51 +340,40 @@ func (a *Agent) StartAPI() {
 		if err := a.api.Shutdown(ctx); err != nil {
 			logger.WithError(err).Error("error shutting down the API server")
 		}
-
-		a.wg.Done()
 	}()
 }
 
 // StartSocketListeners starts the agent's TCP and UDP socket listeners.
-func (a *Agent) StartSocketListeners() {
-	if _, _, err := a.createListenSockets(); err != nil {
+func (a *Agent) StartSocketListeners(ctx context.Context) {
+	if _, _, err := a.createListenSockets(ctx); err != nil {
 		logger.WithError(err).Error("unable to start socket listeners")
 	}
 }
 
-// Stop shuts down the agent. It will block until all listening goroutines
-// have returned.
-func (a *Agent) Stop() {
-	a.cancel()
-	close(a.stopping)
-	if err := a.apiQueue.Close(); err != nil {
-		logger.WithError(err).Error("error closing API queue")
-	}
-	a.wg.Wait()
-}
-
 // StartStatsd starts up a StatsD listener on the agent, logs an error for any
 // failures.
-func (a *Agent) StartStatsd() {
+func (a *Agent) StartStatsd(ctx context.Context) {
 	logger.Info("starting statsd server on address: ", a.statsdServer.MetricsAddr)
 
 	go func() {
-		if err := a.statsdServer.Run(a.context); err != nil {
+		if err := a.statsdServer.Run(ctx); err != nil && err != ctx.Err() {
 			logger.WithError(err).Errorf("error with statsd server on address: %s, statsd listener will not run", a.statsdServer.MetricsAddr)
 		}
 	}()
 }
 
-func connectWithBackoff(url string, tlsOpts *v2.TLSOptions, header http.Header) transport.Transport {
+func connectWithBackoff(ctx context.Context, selector BackendSelector, tlsOpts *v2.TLSOptions, header http.Header) (transport.Transport, error) {
 	var conn transport.Transport
 
 	backoff := retry.ExponentialBackoff{
 		InitialDelayInterval: 10 * time.Millisecond,
 		MaxDelayInterval:     10 * time.Second,
 		Multiplier:           10,
+		Ctx:                  ctx,
 	}
 
-	if err := backoff.Retry(func(retry int) (bool, error) {
+	err := backoff.Retry(func(retry int) (bool, error) {
+		url := selector.Select()
 		c, err := transport.Connect(url, tlsOpts, header)
 		if err != nil {
 			logger.WithError(err).Error("reconnection attempt failed")
```

Comment:
Is returning `false, nil` intentional here, or would we rather return `false, err`?

Reply:
A quirk in our retry library means that we cannot return non-nil error without terminating the retry loop.

https://github.com/sensu/sensu-go/blob/master/util/retry/retry.go#L74

Reply:
Oh interesting.. thanks for clarifying!